projects/plugins/jetpack/extensions/blocks/premium-content/_inc/subscription-service/class-abstract-token-subscription-service.php

```diff
@@ -481,8 +477,9 @@ public static function clear_token_cookie( $cookie_domain = '' ) {
 			unset( $_COOKIE[ self::JWT_AUTH_TOKEN_COOKIE_NAME ] );
 			setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, '', 1, '/', COOKIE_DOMAIN, is_ssl(), true );
 
-			if ( ! empty( $cookie_domain ) ) {
-				setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, '', 1, '/', $cookie_domain, is_ssl(), true );
+			if ( ! empty( COOKIE_DOMAIN ) ) {
+				// On Simple COOKIE_DOMAIN is set to .wordpress.com but we need to remove the cookie from the actual blog hostname.
+				setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, '', 1, '/', '', is_ssl(), true );
```

Comment:
Needed to remove the cookie domain since the cookie removal was not working with it for some reason...

Probably because the modal is setting it without the domain, but I'm not 100% sure.

Reply:
As you are the only oner using this method, maybe it is worth a small cleanup to make more sense out of it: we should ot have to set multiple cookies here anymore.

Reply:
Agree it needs some cleanup and I would love to do it but first I need to wrap my head around all kinds of subscription flows we have. It's not the case at the moment.

Why do we have two `setcookie` for cleaning it out?

My understanding is, we need:
```php
setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, '', 1, '/', COOKIE_DOMAIN, is_ssl(), true );
```
 because it's the opposite of what we have in the `set_token_cookie` function:
```php
setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, $token, strtotime( '+1 month' ), '/', COOKIE_DOMAIN, is_ssl(), false );
```
This sometimes is not enough, for example for Simple sites the `COOKIE_DOMAIN` is set to `.wordpress.com` but the cookie is set for the actual blog hostname (on frontend by modal?), that's why we need a second `setcookie` to cover this case (see the comment I added there).

I'll be iterating on cleaning this function up as my understanding of subscription flow increases.

Happy to hear your thoughts.

Reply:
Maybe instead of using the `COOKIE_DOMAIN` we should consider using sth like this?

https://github.com/Automattic/jetpack/blob/d60c4002aa778195381257f3986a6913165dea8a/projects/plugins/jetpack/modules/wordads/php/class-wordads-consent-management-provider.php#L105-L119

Reply:
> 				// On Simple COOKIE_DOMAIN is set to .wordpress.com but we need to remove the cookie from the actual blog hostname.

For clarity: I tried both flows of setting the cookie on simple site (front+back), and both set it to the site primary domain:

<img width="400" alt="Screenshot 2024-03-19 at 11 49 05" src="https://github.com/Automattic/jetpack/assets/790558/bcdb95cd-2894-4079-a341-35736a42eb55">
<img width="400" alt="Screenshot 2024-03-19 at 12 11 18" src="https://github.com/Automattic/jetpack/assets/790558/b1d5c490-6c52-497b-b291-5377c11dfa78">



So it should be safe to remove the previous `setcookie()` and keep only yours:
```
setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, '', 1, '/', COOKIE_DOMAIN, is_ssl(), true );
```

Reply:
> Maybe instead of using the `COOKIE_DOMAIN` we should consider using sth like this?
> 
> https://github.com/Automattic/jetpack/blob/d60c4002aa778195381257f3986a6913165dea8a/projects/plugins/jetpack/modules/wordads/php/class-wordads-consent-management-provider.php#L105-L119

Do you have a flow where you don't get the actual full domain?

Reply:
I log in through the Subscribe Block and log out by using the Subscriber Login logout link. This is what I get on logout in `clear_token_cookie` function.
1. On Jurassic Tube the `COOKIE_DOMAIN` is set to `false`

```
[19-Mar-2024 13:29:23 UTC] [c266-3042 => pkuliga.jurassic.tube/wp-login.php?action=logout&redirect_to=https%3A%2F%2Fpkuliga.jurassic.tube%2F2024%2F02%2F14%2Fhello-world%2F&_wpnonce=a922deb4c8]
[19-Mar-2024 13:29:23 UTC] [c266-3042] COOKIE_DOMAIN
[19-Mar-2024 13:29:23 UTC] [c266-3042] FALSE
```

2. On Simple sites it's set to `.wordpress.com`

```
[19-Mar-2024 13:33:08 UTC] [0edc-2695977 => pkuliga.wordpress.com/wp-login.php?action=logout&redirect_to=https%3A%2F%2Fpkuliga.wordpress.com%2F2024%2F02%2F06%2Fnew-shiny-post%2F&_wpnonce=48e73bf26e]
[19-Mar-2024 13:33:08 UTC] [0edc-2695977] COOKIE_DOMAIN
[19-Mar-2024 13:33:08 UTC] [0edc-2695977] .wordpress.com
```

Reply:
@pkuliga I meant when we set the cookie, not when we clear it. 

It just seems like `clear_token_cookie` is clearly outdated and not aligned anymore with latest changes., and should be:
```
public static function clear_token_cookie() {
   unset( $_COOKIE[ self::JWT_AUTH_TOKEN_COOKIE_NAME ] );
   self::set_token_cookie('');
}
```

Then `set_token_cookie` needs to be made static.

Do you agree on this? Can you implement this approach?

Reply:
It won't work because of `! empty( $token )` in `set_token_cookie`. I'm removing this condition and we'll see.

Reply:
Logging out doesn't work for me on Simple site since the cookie is set for the `pkuliga.wordpress.com` but the set-cookie header removes it from the `.wordpress.com`:
```
Set-Cookie: wp-jp-premium-content-session=deleted; expires=Thu, 01-Jan-1970 00:00:01 GMT; Max-Age=0; path=/; domain=.wordpress.com; secure
```

Reply:
> I'm removing this condition and we'll see.

Thanks for looking into that!

I see. You're right. The cookie on Simple [is always set from the front-end](https://github.com/Automattic/jetpack/blob/2bec08f5519dcb79ff795c59bcd7720b9e86c700/projects/plugins/jetpack/extensions/shared/memberships.js#L117).

If `domain=` is not provided,  `, this defaults to the host portion of the current document location and the cookie is not available on subdomains` ([source](https://developer.mozilla.org/en-US/docs/Web/API/Document/cookie#write_a_new_cookie))

You approach makes sense, but I would like to use `set_token_cookie` to keep the code clear enough. 

Using `get_cookie_domain()` is not gonna work either.

As much as the easiest choice might seem to set the cookie to `.wordpress.com` on the front-end, it will be an issue as subscriptions in the token are only for a specific subdomain, and therefore the token will not work on other WPCOM subdomains.

My approach

- Keep this
```
public static function clear_token_cookie() {
   unset( $_COOKIE[ self::JWT_AUTH_TOKEN_COOKIE_NAME ] );
   self::set_token_cookie('');
}
```

- Modify `set_token_cookie` to:
 - allow empty token as you did (maybe we should move the `unset($_COOKIE)` part there if the token is empty?
 - set the domain depending on the platform
 
 ```
 	static function set_token_cookie( $token ) {
		if ( defined( 'TESTING_IN_JETPACK' ) && TESTING_IN_JETPACK ) {
			return;
		}
              
               $domain  = is_wpcom() ? get_blog_details()->domain : COOKIE_DOMAIN;
              
                if( empty( $token ) ) {
                    unset($_COOKIE)
                }


		if (  false === headers_sent() ) {
			// phpcs:ignore Jetpack.Functions.SetCookie.FoundNonHTTPOnlyFalse
			setcookie( self::JWT_AUTH_TOKEN_COOKIE_NAME, $token, strtotime( '+1 month' ), '/', COOKIE_DOMAIN, is_ssl(), false );
		}
	}
 ```

I know it might not look the best, but I think it might be the most balanced approach. After all, COOKIE_DOMAIN is meant to work across all WPCOM simple sites (like login cookie, etc...), while our cookie is restricted to a specific subdomain.

What do you think?

Reply:
Adding `get_primary_redirect()` will also support custom domains.

```
$domain  = is_wpcom() ? get_primary_redirect() : COOKIE_DOMAIN;
```

Reply:
It makes sense, I'll explore this solution tomorrow.

> I see. You're right. The cookie on Simple [is always set from the front-end](https://github.com/Automattic/jetpack/blob/2bec08f5519dcb79ff795c59bcd7720b9e86c700/projects/plugins/jetpack/extensions/shared/memberships.js#L117).

If this is the case I would love to make things clear for others and:
1. Add a comment with an explanation to the `set_token_cookie` that setting the cookie on the backend takes place only for non Simple sites.
2. Update [this comment](https://github.com/Automattic/jetpack/blob/fa689815f95a9fb0710bb418b620b4ff4c08b0d7/projects/plugins/jetpack/extensions/shared/memberships.js#L113-L114) as well since we're not setting the cookie "just in case". It's actually needed for Simple sites for whatever reason.

WDYT?

Reply:
> Update [this comment](https://github.com/Automattic/jetpack/blob/fa689815f95a9fb0710bb418b620b4ff4c08b0d7/projects/plugins/jetpack/extensions/shared/memberships.js#L113-L114) as well since we're not setting the cookie "just in case". It's actually needed for Simple sites for whatever reason.

You are right. This comment made sense before when the token was set in the URL before reloading the page, and the token was set from the URL. This flow does not happen anymore when coming from the modal.


> Add a comment with an explanation to the set_token_cookie that setting the cookie on the backend takes place only for non Simple sites.

This is not needed, though. It is not used in a normal flow, but we do use it for debugging purposes.

Reply:
> This is not needed, though. It is not used in a normal flow, but we do use it for debugging purposes.

Not sure about it. In my testing, it's used on a Jurassic Tube site, when subscribing through the Subscribe block and clicking the "Confirm email" button in the email.

Reply:
Ok, the cleanup is done. I decided to:

1. Do not set the cookie domain at all and rely on default behavior. This simplifies things.
2. Do not call `set_token_cookie` from `clear_token_cookie`. Coupling those functions is not needed and IMO introduces extra complexity. Let's keep them separated.

Tested it on Jurassic Tube and Simple and it works as expected.

Reply:
> Not sure about it. In my testing, it's used on a Jurassic Tube site, when subscribing through the Subscribe block and clicking the "Confirm email" button in the email.

The comment was about adding a comment that "setting the cookie on the backend takes place only for non Simple sites.".
I was saying that setting the cookie in the back-end is used:
- for non-Simple sites in the "regular" user flow  (as you experienced with your JT, for example)
- for simple sites, but when debugging (to set the token)

Reply:
> Do not set the cookie domain at all and rely on default behavior. This simplifies things.

Agreed.

Reply:
> Do not call set_token_cookie from clear_token_cookie. Coupling those functions is not needed and IMO introduces extra complexity. Let's keep them separated.

From what I see, you have now 2 different method that do exactly the same thing (one return case is not explicit in ), expect one sets an empty string and the other sets a string. I understand it can seem less complex, but it will can make maintenance more difficult and is error-prone (if we were to change some parameters in `set_cookie` for example).
 
But I'll leave that decision to you.

Reply:
What about removing `clear_token_cookie` altogether and call `set_token_cookie("")` (and removing the unecessary empty token check)? That might be the right approach...

Reply:
`set_token_cookie("")` is a bit too unclear at first sight. What this function is even doing? Does it set the cookie value to "" or remove the cookie completely? I prefer staying with the `clear_token_cookie` as it's more descriptive.

@millerf thanks for your patience and valuable comments. We managed to clean up and simplify things a bit. My understanding of the subscription flows has increased as well!

Hopefully, we won't introduce any regression 😅

Reply:
> Hopefully, we won't introduce any regression 😅

You set the domain to "". It is unclear from the documentation what exact domain will be set (main domain or subdomain). This is a possible regression needs testing.